### ISSUE TITLE
Clang does not honor __builtin_expect and [[likely]]/[[unlikely]] when used with goto in a block-statement.

### ISSUE THREAD

**Issue by EAirPeter** [llvm/llvm-project#53308]
#### Short version
See the sample code I composed in [Compiler Explorer](https://gcc.godbolt.org/z/P9coaKrx9).

#### Long version
It seems that clang does not generate code as _expected_ for the following pattern:
```cpp
void test()
{
  {
    bool cond = foo();
    if (__builtin_expect(foo(), 1)) // or [[likely]]
      goto jGoto;
  }
  // bar1 ...
jGoto:
  // bar2 ...
}
```
The word _expected_ here means to emit the desired branch (`bar2`) such that it is less likely to suffer from branch prediction penalty. On x86, it should place `bar2` at branch-not-taken path in x86 (right after a Jcc instruction, as shown [here](https://gcc.godbolt.org/z/P9coaKrx9)).

A weird fact I realized is that Clang does generate code as _expected_ for the pattern without the surrounding block:
```cpp
void test()
{
  bool cond = foo();
  if (__builtin_expect(foo(), 1)) // or [[likely]]
    goto jGoto;
  // bar1 ...
jGoto:
  // bar2 ...
}
```

The inconsistent behavior of the compiler our production code, hopefully LLVM can fix it.

Note: I also tested this with GCC, it emits identical code for both pattern.

**Comment by ecnelises** [comment]
Yes, in your case (the branch is outside of `cond`'s scope), clang doesn't even generate the branch weight information in IR (expecting `!3 = !{!"branch_weights", i32 2000, i32 1}`).

Moving `cond` outside of the scope is a workaround.
